intrastat_product/models/account_move.py

```diff
@@ -90,17 +90,61 @@ def _get_intrastat_line_vals(self, line):
             if not hs_code:
                 return vals
             weight, qty = decl_model._get_weight_and_supplunits(line, hs_code)
+            if not self.env.context.get("bypass_transaction_weight_convert"):
+                weight = int(weight)
             vals.update(
                 {
                     "invoice_line_id": line.id,
                     "hs_code_id": hs_code.id,
-                    "transaction_weight": int(weight),
+                    "transaction_weight": weight,
                     "transaction_suppl_unit_qty": qty,
                     "product_origin_country_id": line.product_id.origin_country_id.id,
                 }
             )
         return vals
 
+    def _prepare_intrastat_line_info(self, line):
+        is_intrastat_line = bool(line._name == "account.move.intrastat.line")
+        product = line.product_id
+        _self = self.with_context(bypass_transaction_weight_convert=True)
+        return {
+            "product_id": product,
+            "hs_code_id": (
+                line.hs_code_id if is_intrastat_line else product.hs_code_id
+            ),
+            "weight": (
+                line.transaction_weight
+                if is_intrastat_line
+                else _self._get_intrastat_line_vals(line)["transaction_weight"]
+            ),
+            "origin_country_id": (
+                line.product_origin_country_id
+                if is_intrastat_line
+                else product.origin_country_id
+            ),
+        }
+
+    def _get_intrastat_lines_info(self):
+        """We obtain a list of information that we will need to group at the end by
+        product and sum weight.
+        """
+        items = (
+            self.invoice_line_ids.filtered(
+                lambda x: x.product_id.hs_code_id and x.product_id.origin_country_id
+            )
+            if not self.intrastat_line_ids
+            else self.intrastat_line_ids
+        )
+        res = {}
+        for line in items:
+            res.setdefault(line.product_id.id, {"weight": 0})
+            vals = self._prepare_intrastat_line_info(line)
+            weight = vals["weight"]
+            del vals["weight"]
```

Comment:
Not sure why you need this, but in any case, it can be done as:

```suggestion
            weight = vals.pop("weight")
```